src/middleware.ts

```diff
@@ -248,7 +248,12 @@ export const persist = <S extends State>(
       blacklist.forEach((key) => delete state[key])
     }
 
-    return storage?.setItem(name, await serialize({ state, version }))
+    if (storage) {
+      const serializeResult = serialize({ state, version })
+      return serializeResult instanceof Promise
+        ? serializeResult.then((value) => storage.setItem(name, value))
+        : storage.setItem(name, serializeResult)
```

Comment:
I think your idea of `toThenable()` is super neat.
```js
const toThenable = (fn, input) => {
  try {
    const result = fn(input)
    if (result instanceof Promise) {
      return result
    }
    return { then: (onFulfilled) => onFulfilled(result) }
  } catch (e) {
    return { then: (onFulfilled, onRejected) => onRejected(e) }
  }
}
```

```suggestion
      toThenable(serialize, { state, version }).then(
        (value) => storage.setItem(name, value),
        (err) => { throw e }
      )
```

Something like this? @TheEhsanSarshar

Reply:
Yeah, you did it. I was thinking on how to implement it. Thank you @dai-shi

Reply:
Good. Still more to do, including to make it TypeScript.